src/Equinox.MessageDb/MessageDb.fs

```diff
@@ -206,7 +206,7 @@ module Read =
             let batchLog = log |> Log.prop "batchIndex" batchCount
             let! slice = readSlice pos batchCount batchLog ct
             version <- max version slice.LastVersion
-            result.AddRange(slice.Messages)
+            state <- slice.Messages |> Array.chooseV tryDecode |> fold state
```

Comment:
seq.chooseV - contract does not require it to be an array (was putting it into an array for best memory usage and to be sure the db logic can release stuff etc)

Reply:
> contract does not require it to be an array

Should it? It'd make the calling code not need the explicit type annotations
```fsharp
let fold = Array.fold evolve
// vs
let fold: State -> Event seq -> State = Seq.fold evolve
```

```suggestion
            state <- slice.Messages |> Seq.chooseV tryDecode |> fold state
```

Reply:
If I wanted to convey the concept that you have the choice of folding them in groups to optimise, the array makes sense

The actual reason to drop it to seq is twofold
- let people do manipulations (upconversions etc, without having to create interstitial arrays)
- I'm not 100% sure but in general in F# its that tiny bit easier to use `list` in tests in some cases, and having the interface be seq makes that more possible

I'll grant you that the Seq.fold evolve syntax thing is a pain (I have not looked for a better solution than the type annotation for years though)
does `let fold s = Seq.fold evolve s` work (not saying it's necessarily better though)?

But there may be swings and roundabouts in this, e.g. there may be common stores that force it to arrays all the time so on balance it might make things simpler - we're not V4 yet if you're up for the spiking (but I would say only 40% chance of a good result)!

Reply:
> If I wanted to convey the concept that you have the choice of folding them in groups to optimise, the array makes sense

A common pattern I have in my JS applications is

```ts
const fold = (state, events) => events.length ? last(events).data : state
```
For streams where the state can be reconstructed from the latest event. Using a seq would imply this is unwise because you have to iterate the entire seq to get the last element

Reply:
I did once have special case behavior for AnyKnownState mode and went down that rabbit hole
The load will only load 0/one when configured that way, so a `Seq.tryExactlyOne` would be an equivalently pleasing semantic, no?
I think I had some Seq.tryLast stuff too. Can't recall
The job here is to fold though and you should not be called with empty events
Does the fold function accepted by AnyKnownState even give you the state to fold from (it shouldnt)

Reply:
For now best to do a Seq.choose and leave the array change for later